nexus/src/app/sagas/instance_create.rs

```diff
@@ -60,174 +50,286 @@ pub struct Params {
+            parent_builder: &mut steno::DagBuilder,
+            params: S,
+            which: usize,
+        ) -> Result<(), SagaInitError> {
+            // The "parameter" node is a constant node that goes into the outer
+            // saga.  Its value becomes the parameters for the one-node subsaga
+            // (defined below) that actually creates each NIC.
+            let params_node_name = format!("{}_params{}", node_basename, which);
+            parent_builder.append(Node::constant(
+                &params_node_name,
+                serde_json::to_value(&params).map_err(|e| {
+                    SagaInitError::SerializeError(params_node_name.clone(), e)
+                })?,
+            ));
 
-    // Similar to disks (see block comment below), create a sequence of
-    // action-undo pairs for each requested external IP, to make sure we always
-    // unwind after failures.
-    for i in 0..MAX_EXTERNAL_IPS_PER_INSTANCE {
-        template_builder.append(
-            format!("external_ip_id{i}").as_str(),
-            "CreateExternalIpId",
-            new_action_noop_undo(saga_generate_uuid),
-        );
+            let output_name = format!("{}{}", node_basename, which);
+            parent_builder.append(Node::subsaga(
+                output_name.as_str(),
+                subsaga_builder.build()?,
+                params_node_name,
+            ));
+            Ok(())
+        }
 
-        template_builder.append(
-            format!("external_ip{i}").as_str(),
-            "CreateExternalIp",
-            ActionFunc::new_action(
-                async move |sagactx| {
-                    sic_allocate_instance_external_ip(sagactx, i).await
-                },
-                async move |sagactx| {
-                    sic_allocate_instance_external_ip_undo(sagactx, i).await
-                },
-            ),
-        );
-    }
+        // We use a similar pattern here for NICs, external IPs and disks.  We
+        // want one saga action per item (i.e., per NIC, per external IP, or per
+        // disk).  That makes it much easier to make actions and undo actions
+        // idempotent.  But the user may ask for a variable number of these
+        // items.  Previous versions of Steno required the saga DAG to be fixed
+        // for all runs of a saga.  To address this, we put a static limit on
+        // the number of NICs, external IPs, or disks that you can request.
+        // Here where we're building the saga DAG, we always add that maximum
+        // number of nodes and we just have the extra nodes do nothing.
+        //
+        // An easy way to pass this kind of information to an action node is to
+        // wrap it in a subsaga and put that information into the subsaga
+        // parameters.  That's what we do below.  subsaga_append() (defined
+        // above) handles much of the details.
+        //
+        // TODO-cleanup More recent Steno versions support more flexibility
```

Comment:
Do we want to fix this here, or leave it to a follow up PR?

Reply:
I considered fixing this here but the surgery got more complicated than I wanted to take on here.  (I think you want to split the create-nic action into two separate actions: one for the default config and one for custom.  Both would have the same undo function.)

Reply:
Ok, that makes sense. I think this works as a first cut that enables the same behavior with steno 0.2.